### PR TITLE
Implement mfs API

### PR DESCRIPTION
**Pull request by whyrusleeping**
WIP api changes to implement mfs.

License: MIT
Signed-off-by: Jeromy <jeromyj@gmail.com>

### CODE REVIEW

**Comment by jbenet** [comment]
ping me when i should take a look

**Comment by whyrusleeping** [comment]
@jbenet you could probably take a look at this when you get a chance.

File: unixfs/format.go [hunk 1]
```diff
@@ -67,6 +67,7 @@ func WrapData(b []byte) []byte {
 	typ := pb.Data_Raw
 	pbdata.Data = b
 	pbdata.Type = &typ
+	pbdata.Filesize = proto.Uint64(uint64(len(b)))
```
Comment (whyrusleeping) [inline unixfs/format.go]: This was a bug.

Comment (jbenet) [inline unixfs/format.go]: :+1:

File: core/commands/mfs.go [hunk 2]
```diff
@@ -0,0 +1,818 @@
...
+}
+
+var MfsCreateCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Create a new mutable filesystem",
+		ShortDescription: `
+Creates a new mutable filesystem based on an optional root hash.
+
+Currently, it creates a filesystem with no special publish actions,
+all changes are merely propogated to the root, and are reflected in the hash
```
Comment (jbenet) [inline core/commands/mfs.go]: propagated

File: mfs/system.go [hunk 3]
```diff
@@ -1,25 +1,22 @@
...
+// package mfs implements an in memory model of a mutable ipns filesystem,
 // to be used by the fuse filesystem.
 //
 // It consists of four main structs:
 // 1) The Filesystem
 //        The filesystem serves as a container and entry point for the ipns filesystem
-// 2) KeyRoots
-//        KeyRoots represent the root of the keyspace controlled by a given keypair
+// 2) Root
+//        Root represent the root of the keyspace controlled by a given keypair
```
Comment (jbenet) [inline mfs/system.go]: This comment should not say "the root of the keyspace controlled by a given keypair" any more, right?

File: unixfs/format.go (hunk 1 above)
Comment (jbenet) [inline unixfs/format.go]: :+1:

File: unixfs/mod/dagmodifier_test.go [hunk 4]
```diff
@@ -416,6 +416,15 @@ func TestDagTruncate(t *testing.T) {
...
 	}
 
+	size, err := dagmod.Size()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if size != 12345 {
+		t.Fatal("size was incorrect!")
+	}
```
Comment (jbenet) [inline unixfs/mod/dagmodifier_test.go]: :+1: yay tests

File: unixfs/mod/dagmodifier_test.go [hunk 5]
```diff
@@ -429,6 +438,21 @@ func TestDagTruncate(t *testing.T) {
...
+
+	size, err = dagmod.Size()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if size != 10 {
+		t.Fatal("size was incorrect!")
+	}
```
Comment (jbenet) [inline unixfs/mod/dagmodifier_test.go]: :+1: yay tests

**Comment by jbenet** [comment]
Skimmed it, looks right. I need to sink into it + play with it a bit. not sure i can before next week. if priority, i can do it, but have other pressing things at the moment.

File: mfs/system.go (hunk 3 above)
Comment (whyrusleeping) [inline mfs/system.go]: very good point

**Comment by whyrusleeping** [comment]
@jbenet no big hurry. Do your things.

**Comment by jbenet** [comment]
I already need this for some improvements to add :) :+1:

File: core/commands/mfs.go [hunk 6]
```diff
@@ -0,0 +1,819 @@
...
+	mfs "github.com/ipfs/go-ipfs/mfs"
+	path "github.com/ipfs/go-ipfs/path"
+	ft "github.com/ipfs/go-ipfs/unixfs"
+
+	context "github.com/ipfs/go-ipfs/Godeps/_workspace/src/golang.org/x/net/context"
+)
+
+type mfsMountListing struct {
+	Mounts []mfs.RootListing
+}
```
Comment (jbenet) [inline core/commands/mfs.go]: can we scope all the mfs commands into its own dir + separate files (file per cmd)? this package is massive and it will help break things up later.

File: core/commands/mfs.go [hunk 7]
```diff
@@ -0,0 +1,819 @@
...
+They can be seeded with an initial root hash, or by default are an empty directory.
+
+NOTICE:
+This API is currently experimental, likely to change, and may potentially be
+unstable. This notice will be removed when that is no longer the case. Feedback
+on how this API could be improved is very welcome.
+		`,
+	},
+	Options: []cmds.Option{
+		cmds.StringOption("s", "session", "the name of the filesystem to operate on (default=local)"),
```
Comment (jbenet) [inline core/commands/mfs.go]: i think `"name of the editing session"` is clearer. naming a filesystem is a bit confusing. Also pls explain some about sessions + naming them in the `ShortDescription` -- even if WIP, doc helps a lot

File: core/commands/mfs.go [hunk 8]
```diff
@@ -0,0 +1,819 @@
...
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+	},
+}
+
+var MfsCloseCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline:          "Close an open filesystem",
```
Comment (jbenet) [inline core/commands/mfs.go]: not sure the Open/Close metaphor fits well?  Start/End sessions is more ok. but still odd.

File: core/commands/mfs.go [hunk 9]
```diff
@@ -0,0 +1,819 @@
...
+		}
+
+		s, found, err := req.Option("session").String()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+		if !found {
+			s = "local"
+		}
```
Comment (jbenet) [inline core/commands/mfs.go]: this section (265-272) should be a sep func

File: core/commands/mfs.go [hunk 10]
```diff
@@ -0,0 +1,819 @@
...
+			return
+		}
+
+		root, err := node.Mfs.GetRoot(sess)
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		rdir := root.GetValue().(*mfs.Directory)
```
Comment (jbenet) [inline core/commands/mfs.go]: there's a lot of boilerplate in these cmds that could be abstracted out?

File: unixfs/format.go (hunk 1 above)
Comment (jbenet) [inline unixfs/format.go]: may want to just push this earlier, so it's no longer a bug, while we land this

File: core/commands/mfs.go (hunk 8 above)
Comment (whyrusleeping) [inline core/commands/mfs.go]: I was thinking `Mount/Unmount`

**Comment by whyrusleeping** [comment]
I need to figure out the test failures still, but some CR would be good.

File: core/commands/mfs/mfs.go [hunk 11]
```diff
@@ -0,0 +1,689 @@
...
+Mfs is an API for manipulating ipfs objects as if they were a unix filesystem.
+They can be seeded with an initial root hash, or by default are an empty directory.
+
+'Sessions' may be created with the create command. Once created, they show up in
+the output of the base mfs command.
+
+NOTICE:
+This API is currently experimental, likely to change, and may potentially be
+unstable. This notice will be removed when that is no longer the case. Feedback
+on how this API could be improved is very welcome.
```
Comment (jbenet) [inline core/commands/mfs/mfs.go]: provide a link to an issue in which to provide feedback

File: core/commands/mfs/mfs.go [hunk 12]
```diff
@@ -0,0 +1,689 @@
...
+var MfsCreateCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Create a new mutable filesystem",
+		ShortDescription: `
+Creates a new mutable filesystem based on an optional root hash.
+
+Currently, it creates a filesystem with no special publish actions,
+all changes are merely propagated to the root, and are reflected in the hash
+displayed by 'ipfs mfs'. In the future, this command will allow you to specify
+publish actions and close actions for a given filesystem
```
Comment (jbenet) [inline core/commands/mfs/mfs.go]: what do you mean by `publish` or `close`? i think this is confusing. maybe remove those mentions.

File: core/commands/mfs/mfs.go [hunk 13]
```diff
@@ -0,0 +1,689 @@
...
+			}
+			return buf, nil
+		},
+	},
+	Type: mfsMountListing{},
+}
+
+var MfsCreateCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Create a new mutable filesystem",
```
Comment (jbenet) [inline core/commands/mfs/mfs.go]: Hmm not sure this should be called "a filesystem". It's confusing to think about "using ipfs  -- a distributed file system -- to then have to create other filesystems-- what?" (i know it's technically a filesystem, but the overloading is hard on the user)

Maybe this should be called "a mutable root" or "a mutable space", or just "a mutable dag"?

Another option is to, instead of calling it a thing and letting users create roots, maybe we should:

- have a proper root object that links to all these "roots"
- users mutate _that_ root always, and they get only one
- in git-land, consider a repo. a repo is really just a "tree" with links, whose names are "branches" and values are "commit" hashes.
- likewise here we could have the root be the user's "local" data, every root of an "ipfs add" gets added to a node under that called "/local/added" (like a log) (( maybe this is exactly what already happens )) 
- and "mfs create", would be the equivalent of "ipfs local link <name> <root>"  ?

---

- we could also name these  "branches" (bad if we want to do proper commit branches later) or 
- "views" (like a "view on the data")

File: core/commands/mfs/mfs.go [hunk 14]
```diff
@@ -0,0 +1,689 @@
...
+		Tagline: "import a given file into a filesystem",
+		ShortDescription: `
+Mfs 'put' takes a file that already exists in ipfs and imports it into a
+given mfs filesystem.
+		`,
+	},
+
+	Arguments: []cmds.Argument{
+		cmds.StringArg("object", true, false, "ipfspath of object to import"),
+		cmds.StringArg("path", true, false, "path within filesystem to import to"),
```
Comment (jbenet) [inline core/commands/mfs/mfs.go]: maybe can detect whether it's an ipfs path with `/ipfs/` as the prefix? (get users thinking about `/ipfs/...` and their local fs coexisting)

File: core/commands/mfs/mfs.go [hunk 15]
```diff
@@ -0,0 +1,689 @@
...
+
+var MfsMvCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Move files",
+		ShortDescription: `
+Move files around. Just like traditional unix mv.
+
+Example:
+
+    ipfs mfs -s myfs mv /a/b/c /foo/newc
```
Comment (jbenet) [inline core/commands/mfs/mfs.go]: if we go with the discussion above, maybe can avoid the session with:
```
ipfs mfs mv /myfs/a/b/c /myfs/foo/newc
ipfs mfs mv /local/myfs/a/b/c /local/myfs/foo/newc
ipfs local mv /myfs/a/b/c /myfs/foo/newc
```

File: core/commands/mfs/mfs.go [hunk 16]
```diff
@@ -0,0 +1,689 @@
...
+		ShortDescription: `
+Write data to a file in a given filesystem. This command allows you to specify
+a beginning offset to write to. The entire length of the input will be written.
+
+If the '--create' option is specified, the file will be create if it does not
+exist. Nonexistant intermediate directories will not be created.
+
+Example:
+
+	echo "hello world" | ipfs mfs -s myfs --create /a/b/file
```
Comment (jbenet) [inline core/commands/mfs/mfs.go]: maybe
```
echo "hello world" | ipfs mfs --create /myfs/a/b/file
echo "hello world" | ipfs mfs write --create /myfs/a/b/file
echo "hello world" | ipfs mfs write --truncate /myfs/a/b/file
```

**Comment by jbenet** [comment]
- some comments above, mainly about UX.
- will play around with the current UX some more
- i haven't read all the code super closely, will do so in next day or two

File: core/commands/mfs/mfs.go (hunk 13 above)
Comment (whyrusleeping) [inline core/commands/mfs/mfs.go]: I'm okay with changing it to something like mutable space, but then what would we rename the command to?

File: core/commands/mfs/mfs.go (hunk 14 above)
Comment (whyrusleeping) [inline core/commands/mfs/mfs.go]: i was thinking of having a `cmds.IpfsPathArg` that would validate that the given argument was an ipfs path.

File: core/commands/mfs/mfs.go (hunk 15 above)
Comment (whyrusleeping) [inline core/commands/mfs/mfs.go]: I like the idea of paths instead of a session var, although when dealing with the api, i've found the session flag to be easier. Less path manipulation to worry about.

File: core/commands/mfs/mfs.go (hunk 13 above)
Comment (whyrusleeping) [inline core/commands/mfs/mfs.go]: also, not super sold yet on the idea of having a `/local/` we lose out on this being a more generic thing.. but maybe thats okay? idk

**Comment by whyrusleeping** [comment]
@jbenet after thinking about it a bit i would be okay with `/session/path/to/thing` over using the session flag. sound good to you?

**Comment by whyrusleeping** [comment]
although, if we do that, then we have to deal with 'cross filesystem' operations. which gets interesting.

**Comment by whyrusleeping** [comment]
@jbenet alright, now using paths. I'm not sure what circleCI's issue is though: https://circleci.com/gh/ipfs/go-ipfs/1080

File: core/commands/mfs/mfs.go [hunk 17]
```diff
@@ -0,0 +1,638 @@
...
+
+var MfsCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Manipulate mutable filesystems",
+		ShortDescription: `
+Mfs is an API for manipulating ipfs objects as if they were a unix filesystem.
+They can be seeded with an initial root hash, or by default are an empty directory.
+
+'Sessions' may be created with the create command. Once created, they show up in
+the output of the base mfs command.
```
Comment (jbenet) [inline core/commands/mfs/mfs.go]: - [ ] session comment outdated? not clear how sessions work. 
- [ ] need examples in this doc

File: core/commands/mfs/mfs.go [hunk 18]
```diff
@@ -0,0 +1,638 @@
...
+
+NOTICE:
+This API is currently experimental, likely to change, and may potentially be
+unstable. This notice will be removed when that is no longer the case. Feedback
+on how this API could be improved is very welcome on the following issue:
+https://github.com/ipfs/go-ipfs/issues/1607
+		`,
```
Comment (jbenet) [inline core/commands/mfs/mfs.go]: remove the whitespace up to 

    `,

otherwise outputs too much whitespace

**Comment by jbenet** [comment]
- [ ] rebase on current master
- [ ] seems to not work in online mode, and complains about online mode.

- [ ] session comment outdated? not clear how sessions work. https://github.com/ipfs/go-ipfs/pull/1550/files#r38378755
- [ ] need examples on doc https://github.com/ipfs/go-ipfs/pull/1550/files#r38378755
- [ ] whitespace here https://github.com/ipfs/go-ipfs/pull/1550/files#r38378799
- [ ] i cant run this to test it.
- [ ] this thing needs extensive shareness tests, would've caught \o

File: core/commands/mfs/mfs.go [hunk 19]
```diff
@@ -0,0 +1,638 @@
...
+		cmds.Text: func(res cmds.Response) (io.Reader, error) {
+			out := res.Output().(*mfsMountListing)
+			if len(out.Mounts) == 1 && out.Mounts[0].Name == "" {
+				// a session was specified, print out just the hash
+				return strings.NewReader(out.Mounts[0].Hash.B58String()), nil
+			}
+
+			buf := new(bytes.Buffer)
+			for _, o := range out.Mounts {
+				fmt.Fprintf(buf, "%s - %s\n", o.Name, o.Hash)
```
Comment (jbenet) [inline core/commands/mfs/mfs.go]: - [ ] remove the `-` in output

File: core/commands/mfs/mfs.go [hunk 20]
```diff
@@ -0,0 +1,638 @@
...
+	},
+	Type: MfsLsOutput{},
+}
+
+var MfsPutCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "import a given file into a filesystem",
+		ShortDescription: `
+Mfs 'put' takes a file that already exists in ipfs and imports it into a
+given mfs filesystem.
```
Comment (jbenet) [inline core/commands/mfs/mfs.go]: - [ ] example for `ipfs mfs put`

File: core/commands/mfs/mfs.go [hunk 21]
```diff
@@ -0,0 +1,638 @@
...
+		`,
+	},
+
+	Arguments: []cmds.Argument{
+		cmds.StringArg("object", true, false, "ipfspath of object to import"),
```
Comment (jbenet) [inline core/commands/mfs/mfs.go]: - [ ] maybe `ipfs-path`

File: core/commands/mfs/mfs.go (hunk 14 above)
Comment (jbenet) [inline core/commands/mfs/mfs.go]: - [ ] maybe rename it to `target` or `dst`

File: core/commands/mfs/mfs.go (hunk 20 above)
Comment (jbenet) [inline core/commands/mfs/mfs.go]: maybe involving `ipfs add` as a way to add files to ipfs, and then using the hash to `ipfs mfs put` it

File: core/commands/mfs/mfs.go (hunk 13 above)
Comment (jbenet) [inline core/commands/mfs/mfs.go]: benefits of putting all the things under `/local` or `/mfs` or `/files`:
- could have an option to `ipfs add` that links the add root into a location in `/mfs`
- mfs could stand for `managed file system` or something, too.
- could use all the mfs tooling with arbitrary ipfs-paths, which is like a more hardcore, and unixfs aware, ipfs object patch. 
- if mfs is so unixfs based, maybe mfs should just be `ipfs files` ... and combine the prior idea

File: core/commands/mfs/mfs.go [hunk 22]
```diff
@@ -0,0 +1,638 @@
...
+			for _, o := range out.Mounts {
+				fmt.Fprintf(buf, "%s - %s\n", o.Name, o.Hash)
+			}
+			return buf, nil
+		},
+	},
+	Type: mfsMountListing{},
+}
+
+var MfsCreateCmd = &cmds.Command{
```
Comment (jbenet) [inline core/commands/mfs/mfs.go]: - [ ] remove mfs create command. no longer needed i think. can mkdir and create files from paths. create made sense in the "set of sessions" model. not in the "everything has one root" model.

File: core/commands/mfs/mfs.go [hunk 23]
```diff
@@ -0,0 +1,638 @@
...
+			return
+		}
+	},
+}
+
+type Object struct {
+	Hash string
+}
+
+var MfsCloseCmd = &cmds.Command{
```
Comment (jbenet) [inline core/commands/mfs/mfs.go]: i dont think this makes sense anymore. probably just `ipfs mfs rm /<rootname>`. so:
- [ ] remove `ipfs mfs close`

File: core/commands/mfs/mfs.go (hunk 22 above)
Comment (whyrusleeping) [inline core/commands/mfs/mfs.go]: that changes a lot... you cant just do `ipfs mfs mkdir /dir` and i do want to keep the top level mounts as 'special' objects.

File: core/commands/mfs/mfs.go [hunk 24]
```diff
@@ -0,0 +1,638 @@
...
+
+var log = u.Logger("cmds/mfs")
+
+type mfsMountListing struct {
+	Mounts []mfs.RootListing
+}
+
+var MfsCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Manipulate mutable filesystems",
```
Comment (jbenet) [inline core/commands/mfs/mfs.go]: - [ ] maybe just: 
```
manipulate unix files
```

**Comment by jbenet** [comment]
- [ ] remove the `-` in output
- [ ] `ipfs mfs ls /` (and maybe `ipfs mfs ls` too) should work. the `ipfs mfs` listing can be a shortcut for `ipfs mfs ls /`
- [ ] examples for `ipfs mfs put` in doc https://github.com/ipfs/go-ipfs/pull/1550/files#r38382939
- [ ] rename `object` arg to `ipfs-path` or `src` in `ipfs mfs put` https://github.com/ipfs/go-ipfs/pull/1550/files#r38382968
- [ ] rename `path` arg to `target-path` or `dst` in `ipfs mfs put` https://github.com/ipfs/go-ipfs/pull/1550/files#r38383033
- [ ] `Error: lookup 'demo/' failed: Lookup requires rooted path` annoying. maybe should allow non-rooted. take them as relative to mfs root.
- [ ] expected `ipfs mfs cat` to exist. maybe silent alias to `ipfs mfs read` ?
- [ ] weird error: 
```sh
> ipfs mfs write /demo/foo "foo data"
Error: Option name 'c' used multiple times
# wat? i dont even. 
# also i dont think you want me to add the data there?
> ipfs mfs write --help
Error: Option name 'c' used multiple times
# oh fun! broken. tests kill bugs.
```
- [ ] remove `ipfs mfs create` it no longer makes sense https://github.com/ipfs/go-ipfs/pull/1550/files#r38383476
- [ ] remove ipfs mfs close` (same as above) https://github.com/ipfs/go-ipfs/pull/1550/files#r38383536
- [ ] maybe rename to `ipfs files` https://github.com/ipfs/go-ipfs/pull/1550/files#r38383450
- [ ] expected this to work (maybe use `/files` prefix):
```sh
> ipfs mfs ls /demo
LICENSE
foo
> ipfs mfs put /demo/LICENSE /demo/LICENSE2
# maybe want ipfs mfs cp ? 
# or really just make `ipfs mfs put` work with mfs paths too.
# actually, the prefix would help here, too:
# ipfs mfs put /ipfs/QmcvvLwiGH22g3AWm8kWpNcsoksc7qoRpK1VwoR5D2Gkqo/LICENSE /files/demo/license
# ipfs mfs put /files/demo/LICENSE /files/demo/LICENSE2
```
- [ ] `ipfs mfs mkdir root` failed:
```sh
> ipfs mfs mkdir root
Error: no such rootfs
# this is still working on the "create/sessions" abstraction.
# remove the create/close think and just use mkdir and rm on everything, i think 
```
- [ ] `ipfs mfs ls -l` sizes incorrect: 
```sh
> ipfs mfs ls /demo2 -l
beep	QmZaZ684Z7bpgxrHGrSQJL73eehs69yr21cbpE9AHsZmPP	0
boop	QmcvvLwiGH22g3AWm8kWpNcsoksc7qoRpK1VwoR5D2Gkqo	0
```
- [ ] lies and slander
```sh
> ipfs mfs ls /demo2
boop
> ipfs mfs rm -r /demo2
Error: no such rootfs
# again the sessions/single-root dichotomy
```
- [ ] i think we just need to fix the roots thing and this interface:
```
    ipfs mfs rm <path>...        - remove files
    ipfs mfs cp <source> <target> - copy files
    ipfs mfs read <path>         - read a file
    ipfs mfs write <path> <data> - write to a file
    ipfs mfs mv <source> <target>  - move files
    ipfs mfs mkdir <path>        - make directories
    ipfs mfs ls <path>           - list directory contents
```
- [ ] rename `ipfs mfs put` to `ipfs mfs cp`
- [ ] fix `<source> <target>` to match `mv` and `cp` do
- [ ] fix up sub command taglines to follow unix counterparts (did it for you above \o)
- [ ] `ipfs mfs` tagline fix - https://github.com/ipfs/go-ipfs/pull/1550/files#r38383902

**Comment by jbenet** [comment]
@whyrusleeping not clear which concerns you've addressed. tick the boxes above, ideally referencing commits. this CR is huge, and will otherwise be confusing. anyway, i can see most concerns aren't addressed, so not sure why waiting on me for this.

one more:
- [ ] should work offline. https://github.com/ipfs/go-ipfs/commit/5a1616037bcac5632e6281c1ad6aa7f862a30706#commitcomment-13079894

This PR is two things at once. all this mount/unmount stuff, and the merkledag editing/bubbling thing are two different (and individually large!) things. like proper fs, vs the virtual fs that the kernel exposes. All the mounting business is complex. the "files/" abstraction (of a writable merkledag-backed fs that bubbles updates to a root it's hanging on to) is large enough to be its own thing. (this is what IPNS is layered on top of). The "mounting/unmounting" stuff is very complicated as well, and one could see mounting IPNS, the "files/" thing, and the basic "/ipfs/" resolution on top of that.

In short, this is a monster of a PR. it's huge, and we're trying to satisfy two different, large things: virtual file system that allows mounting others, and a basic mdag fs that bubbles up state changes and keeps state in a root. I would suggest splitting it up into those two. let's address the mdag-only fs thing first, and then do the "mounting arbitrary things on top" second. We can assign the "mfs" name to the second thing, and the name "files/" to the first. (or "mfs" to the first, and "fs/vfs" to the second, not sure).

**Comment by whyrusleeping** [comment]
closing as it was broken into two other PRs which are now merged.
